src/ui/castleinternaljoystickslinux.pas

```diff
@@ -145,54 +147,75 @@ procedure TLinuxJoysticksBackend.Poll(const List: TJoystickList;
+              if Axis = JOY_AXIS_Y then
+                Value := -Value;
+              Joystick.State.Axis[ axis ] := Value;
+              if Assigned(EventContainer.OnAxisMove) then EventContainer.OnAxisMove(Joystick, axis, Value);
+            end;
+          JS_EVENT_BUTTON:
+            case event.value of
+              0:
                 begin
-                  Joystick.State.BtnUp[ event.number ] := True;
-                  Joystick.State.BtnPress   [ event.number ] := False;
-                  if Assigned(EventContainer.OnButtonUp) then EventContainer.OnButtonUp(Joystick, event.number);
-                  Joystick.State.BtnCanPress[ event.number ] := True;
-                end;
-
-                Joystick.State.BtnDown[ event.number ] := False;
-              end;
-            1:
-              begin
-                Joystick.State.BtnDown[ event.number ] := True;
-                if Assigned(EventContainer.OnButtonDown) then EventContainer.OnButtonDown(Joystick, event.number);
-                Joystick.State.BtnUp  [ event.number ] := False;
-                if Joystick.State.BtnCanPress[ event.number ] then
+                  if Joystick.State.BtnDown[ event.number ] then
                   begin
-                    Joystick.State.BtnPress   [ event.number ] := True;
-                    if Assigned(EventContainer.OnButtonPress) then EventContainer.OnButtonPress(Joystick, event.number);
-                    Joystick.State.BtnCanPress[ event.number ] := False;
+                    Joystick.State.BtnUp[ event.number ] := True;
+                    Joystick.State.BtnPress   [ event.number ] := False;
+                    if Assigned(EventContainer.OnButtonUp) then EventContainer.OnButtonUp(Joystick, event.number);
+                    Joystick.State.BtnCanPress[ event.number ] := True;
                   end;
-              end;
-          end;
+
+                  Joystick.State.BtnDown[ event.number ] := False;
+                end;
+              1:
+                begin
+                  Joystick.State.BtnDown[ event.number ] := True;
+                  if Assigned(EventContainer.OnButtonDown) then EventContainer.OnButtonDown(Joystick, event.number);
+                  Joystick.State.BtnUp  [ event.number ] := False;
+                  if Joystick.State.BtnCanPress[ event.number ] then
+                    begin
+                      Joystick.State.BtnPress   [ event.number ] := True;
+                      if Assigned(EventContainer.OnButtonPress) then EventContainer.OnButtonPress(Joystick, event.number);
+                      Joystick.State.BtnCanPress[ event.number ] := False;
+                    end;
+                end;
+            end;
+        end;
+        BytesRead := FpRead( BackendInfo.Device, event, 8 );
+      until BytesRead <> 8
+    else
+      if fpgeterrno <> EAGAIN then
+      begin
+        WritelnLog('Joystick error, possibly it was unplugged. Trying to run Joysticks.Initialize.');
```

Comment:
This log is no longer valid, since we call `OnDisconnect` instead of `Joysticks.Initialize` directly. Better and simpler:

```
WritelnLog('Joystick %s was disconnected', [Joystick.Info.Name]);
```